projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/func/LLVMCallNode.java

```diff
@@ -314,36 +295,35 @@ public LLVMFunctionCallChainStartNode(LLVMExpressionNode functionCallNode, LLVMF
 
         @Override
         public Object executeGeneric(VirtualFrame frame) {
-            try {
-                Object[] args = evaluateArgs(frame);
-                TruffleObject function = functionCallNode.executeTruffleObject(frame);
-                if (function instanceof LLVMFunctionDescriptor) {
-                    return chain.executeDispatch(frame, (LLVMFunctionDescriptor) function, args);
-                } else {
-                    if (foreignExecute == null) {
-                        CompilerDirectives.transferToInterpreterAndInvalidate();
-                        foreignExecute = insert(Message.createExecute(getFunctionArgumentLength(frame)).createNode());
-                    }
-                    return doExecute(frame, foreignExecute, function, args, toLLVM, chain.returnType);
+            Object[] args = evaluateArgs(frame);
+            TruffleObject function = LLVMExpressionNode.expectTruffleObject(functionCallNode, frame);
+            if (function instanceof LLVMFunctionDescriptor) {
+                return chain.executeDispatch(frame, (LLVMFunctionDescriptor) function, args);
+            } else {
+                if (foreignExecute == null) {
+                    CompilerDirectives.transferToInterpreterAndInvalidate();
+                    foreignExecute = insert(Message.createExecute(getFunctionArgumentLength(frame)).createNode());
                 }
-            } catch (UnexpectedResultException e) {
-                CompilerDirectives.transferToInterpreter();
-                throw new IllegalStateException(e);
+                return doExecute(frame, foreignExecute, function, args, toLLVM, chain.returnType);
             }
         }
 
-    }
-
-    public static Object[] convertToPrimitiveArgs(Object[] arguments) {
-        Object[] newArguments = new Object[arguments.length - LLVMCallNode.ARG_START_INDEX];
-        System.arraycopy(arguments, LLVMCallNode.ARG_START_INDEX, newArguments, 0, newArguments.length);
-        CompilerAsserts.compilationConstant(arguments.length);
```

Comment:
I do not think that it is a good idea to replace the `System.arraycopy` calls with the more verbose for loop. The `CompilerAsserts.compilationConstant` ensures that arguments.length is final, so Graal should be able to optimize the copy away.